scanning/disa-alignment/shared.py

```diff
@@ -42,27 +77,121 @@ def disa_scan(host, ds, html, arf):
+        if result == "notselected":
+            continue
+        rule = xccdf_benchmark.find(
+            f".//xccdf:Rule[@id='{full_rule_id}']", nsmap)
+        title = rule.find("xccdf:title", nsmap).text
+        cce_id = rule.find(f"xccdf:ident[@system='{CCE}']", nsmap).text
+        stig_ids = []
+        xpath = f"xccdf:reference[@href='{STIG_REF}']"
+        for stig_ref_el in rule.findall(xpath, nsmap):
+            if stig_ref_el is not None:
+                stig_ids.append(stig_ref_el.text)
+        rule_id = full_rule_id.replace(SSG_RULE_PREFIX, "")
+        sr = SSGRuleResult(rule_id, cce_id, title, stig_ids, result)
+        ssg_results[rule_id] = sr
+    return ssg_results
+
+
+def parse_disa_results(disa_path):
+    disa_results = {}
+    root = ET.parse(disa_path).getroot()
+    xccdf_results = root.find(".//xccdf:TestResult", nsmap)
+    for rule_result in xccdf_results.findall("xccdf:rule-result", nsmap):
+        full_rule_id = rule_result.get("idref")
+        rule_id = full_rule_id.replace(DISA_RULE_PREFIX, "")
+        result = rule_result.find("xccdf:result", nsmap).text
+        disa_results[rule_id] = result
+    return disa_results
+
+
+def compare_result_with_stig_id(result, stig_id, disa_results):
+    if stig_id in disa_results:
+        disa_result = disa_results[stig_id]
+        if result == disa_result:
+            return ComparisonResult.SAME
         else:
-            result_note = f'SSG result: {ssg_result}, DISA result: {disa_result}'
-            results.report('fail', rule_id, result_note)
+            return ComparisonResult.DIFFERENT
+    else:
+        return ComparisonResult.MISSING
+
+
+def compare_results(ssg_results, disa_results):
+    for ssg_rule_result in ssg_results.values():
+        same = 0
+        different = 0
+        missing = 0
+        if len(ssg_rule_result.stig_ids) == 0:
+            ssg_rule_result.final_result = ComparisonResult.MISSING
+        for stig_id in ssg_rule_result.stig_ids:
+            comparison_result = compare_result_with_stig_id(
+                ssg_rule_result.result, stig_id, disa_results)
+            disa_result = disa_results.get(stig_id, "not found")
+            ssg_rule_result.stig_ids_results[stig_id] = disa_result
+            if comparison_result == ComparisonResult.SAME:
+                same += 1
+            elif comparison_result == ComparisonResult.DIFFERENT:
+                different += 1
+            elif comparison_result == ComparisonResult.MISSING:
+                missing += 1
+        if same >= 1:
+            ssg_rule_result.final_result = ComparisonResult.SAME
+        elif different >= 1:
+            ssg_rule_result.final_result = ComparisonResult.DIFFERENT
+        elif missing >= 1:
+            ssg_rule_result.final_result = ComparisonResult.MISSING
+
+
+def get_disa_result_to_str(stig_ids_results):
+    if len(stig_ids_results) == 0:
+        return "Empty"
+    ll = [f"{k}:{v}" for k, v in stig_ids_results.items()]
```

Comment:
To save memory, you don't need to convert everything to a list, only to "destroy" it in the `join()`, you can use generators to let `join()` consume one item at a time.
```suggestion
    ll = (f"{k}:{v}" for k, v in stig_ids_results.items())
```
More on https://wiki.python.org/moin/Generators .